etna/lib/etna/clients/magma/workflows.rb

```diff
@@ -0,0 +1,146 @@
+          actions.each { |a| update.add_action(a) }
+          @target_models = nil
+          target_client.update_model(update)
+        end
+
+        def self.from_api_source(source_project:, source_client:, **kwds)
+          self.new(
+              source_models: source_client.retrieve(RetrievalRequest.new(project_name: source_project, model_name: 'all')).models,
+              **kwds
+          )
+        end
+
+        # Subclass and override when the source <-> target mapping is not 1:1.
+        def target_of_source(model_name)
+          model_name
+        end
+
+        # Subclass and override when the source <-> target attribute mapping is not 1:1.
+        def target_attribute_of_source(model_name, attribute_name)
+          attribute_name
+        end
+
+        # Potentially cyclical, protected aginst re-entry via the seen cache.
+        # Establishes the link attributes in a given model graph.
+        def ensure_model_tree(model_name, seen = Set.new)
+          return unless (source_model = source_models.model(model_name))
+          return if seen.include?(model_name)
+          seen.add(model_name)
+          ensure_model(model_name)
+
+          attributes = source_model.template.attributes
+          attributes.attribute_keys.each do |attribute_name|
+            attribute = attributes.attribute(attribute_name)
+            ensure_model_tree(attribute.link_model_name, seen) unless attribute.link_model_name.nil?
+            ensure_model_attribute(model_name, attribute_name)
+          end
+        end
+
+        def ensure_model_attribute(model_name, attribute_name)
+          return unless (model = source_models.model(model_name))
+          return unless (source_attribute = model.template.attributes.attribute(attribute_name))
+
+          target_model_name = target_of_source(model_name)
+          target_attribute_name = target_attribute_of_source(model_name, attribute_name)
+
+          target_attributes = target_models.model(target_model_name).template.attributes
+          return if target_attributes.attribute_keys.include?(attribute_name)
+
+          add_attribute = AddAttributeAction.new(
+              model_name: target_model_name,
+              attribute_name: target_attribute_name,
+          )
+
+          add_attribute.type = source_attribute.attribute_type
+          add_attribute.description = source_attribute.desc
+          add_attribute.display_name = source_attribute.display_name
+          add_attribute.format_hint = source_attribute.format_hint
+          add_attribute.hidden = source_attribute.hidden
+          add_attribute.link_model_name = source_attribute.link_model_name
+          add_attribute.read_only = source_attribute.read_only
+          add_attribute.unique = source_attribute.unique
+          add_attribute.validation = source_attribute.validation
+          add_attribute.restricted = source_attribute.restricted
```

Comment:
I think I've also seen attribute `options` -- is that something that we would want copied here, too?

Reply:
Good catch.

Reply:
Ah, so `options` and `match` are read only from the json_template; they are actually both derived from the `validation` input which we set.